src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala

```diff
@@ -113,36 +120,59 @@ class TaskReplaceActor(
 
   def handleNewHealthyTask(taskId: String): Unit = {
     healthy += taskId
-    killOldTask(taskId)
+    killNextOldTask(Some(taskId))
     checkFinished()
   }
 
-  def killOldTask(newTaskId: String): Unit = {
+  def killNextOldTask(maybeNewTaskId: Option[String] = None): Unit = {
     if (toKill.nonEmpty) {
-      val killing = toKill.dequeue()
-      log.info(s"Killing old task $killing because $newTaskId became reachable")
-      driver.killTask(buildTaskId(killing))
+      val nextOldTask = toKill.dequeue()
+
+      maybeNewTaskId match {
+        case Some(newTaskId: String) =>
+          log.info(s"Killing old task $nextOldTask because $newTaskId became reachable")
+        case _ =>
+          log.info(s"Killing old task $nextOldTask")
+      }
+
+      outstandingKills += nextOldTask
+      driver.killTask(buildTaskId(nextOldTask))
     }
   }
 
   def checkFinished(): Unit = {
-    if (healthy.size == app.instances) {
-      log.info(s"All tasks for $appId are healthy")
+    if (healthy.size == app.instances && oldTaskIds.isEmpty) {
+      log.info(s"App All new tasks for $appId are healthy and all old tasks have been killed")
+      periodicalRetryKills.cancel()
       promise.success(())
       context.stop(self)
     }
+    else if (log.isDebugEnabled) {
+      log.debug(s"For app: [${app.id}] there are [${healthy.size}] healthy new instances and " +
+        s"[${oldTaskIds.size}] old instances.")
+    }
+  }
+
+  def retryKills(): Unit = {
+    outstandingKills.foreach { id =>
+      log.warning(s"Retrying kill of old task $id")
+      driver.killTask(buildTaskId(id))
+    }
   }
 
   def buildTaskId(id: String): TaskID =
     TaskID.newBuilder()
       .setValue(id)
       .build()
+
+  def scheduleRetryKills(): Unit =
+    periodicalRetryKills = context.system.scheduler.scheduleOnce(15.seconds, self, RetryKills)
 }
 
-private object ReplaceErrorState {
-  def unapply(state: String): Option[String] = state match {
-    case "TASK_ERROR" | "TASK_FAILED" | "TASK_KILLED" | "TASK_LOST" => Some(state)
-    case _ => None
-  }
+object TaskReplaceActor {
+  val TaskFinished = "^TASK_(ERROR|FAILED|FINISHED|LOST|KILLED)$".r
```

Comment:
Rename this property. Irritating that TASK_XXX is TaskFinished where XXX can be FINISHED or something else.